runtime/lua/vim/diagnostic.lua

```diff
@@ -869,12 +873,14 @@ M.handlers.signs = {
     bufnr = get_bufnr(bufnr)
     opts = opts or {}
 
+    if opts.filter then
+      diagnostics = vim.tbl_map(opts.filter, diagnostics)
+    end
```

Comment:
If the `filter` function can modify the diagnostics then they should be copied first (so that they remain immutable).

```suggestion
    if opts.signs and opts.signs.filter then
      diagnostics = vim.tbl_map(opts.signs.filter, vim.deepcopy(diagnostics))
    end
```